### PR TITLE
State table changes

### PR DESCRIPTION
**Pull request by ltratt**
This PR is two vaguely-related changes: https://github.com/softdevteam/lrtable/commit/0f9416664bb9c77fd80db4b95391224ee3f496ab is a performance fix; https://github.com/softdevteam/lrtable/commit/350cfb193937bce86ee27285b5fe57e11fcb133f is an API change. The second commit leads to lots of not-very-interesting mechanical churn, so it's best to look at the commits separately.

### CODE REVIEW

**Comment by snim2** [comment]
In the first commit about speeding up state actions, what `O(?)` does a look-up on a bit-vector have?

**Comment by ltratt** [comment]
It's an array of bits (well, Vector of Bits!) so O(1).

**Comment by ltratt** [comment]
I should have said: this speeds up one of the lrpar recoverers by 10%. Oops!

File: src/lib/statetable.rs [hunk 1]
```diff
@@ -190,36 +195,30 @@ impl StateTable {
...
-            let off = StateTable::actions_offset(self.terms_len, state_idx, TIdx::from(term_idx));
-            if let Some(_) = self.actions.get(&off) {
-                syms.push(Symbol::Term(TIdx::from(term_idx)))
-            }
-        }
-        StateActionIterator{symbols: syms, i: 0}
+    pub fn state_actions(&self, state_idx: StIdx) -> StateActionsIterator {
+        let start_off = usize::from(state_idx) * self.terms_len as usize;
+        StateActionsIterator{stable: &self,
+                             start: start_off,
```
Comment (snim2) [inline src/lib/statetable.rs]: Could simplify this line by calling the variable `start`

File: src/lib/statetable.rs [hunk 2]
```diff
@@ -233,21 +232,27 @@ impl StateTable {
...
+impl<'a> Iterator for StateActionsIterator<'a> {
     type Item = Symbol;
 
     fn next(&mut self) -> Option<Symbol> {
-        if self.i == self.symbols.len() {
-            return None;
+        let mut i = self.i;
+        while i < self.end {
+            if self.stable.used_actions.get(i).unwrap() {
+                self.i = i + 1;
```
Comment (snim2) [inline src/lib/statetable.rs]: Clippy will suggest `self.i += 1;`

File: src/lib/statetable.rs (hunk 1 above)
Comment (ltratt) [reply src/lib/statetable.rs]: True.

**Comment by snim2** [comment]
OK, apart from that very shallow comment, this looks OK to me

**Comment by ltratt** [comment]
Fix pushed. If you're OK with that, should I squash?

File: src/lib/statetable.rs [hunk 3]
```diff
@@ -246,7 +246,7 @@ impl<'a> Iterator for StateActionsIterator<'a> {
         let mut i = self.i;
         while i < self.end {
             if self.stable.used_actions.get(i).unwrap() {
-                self.i = i + 1;
+                self.i += 1;
```
Comment (snim2) [inline src/lib/statetable.rs]: I think my initial comment was nonsense, and this line is now wrong. It was: `self.i = i + 1;` and I think `i` and `self.i` are not always the same -- right?

Comment (ltratt) [reply src/lib/statetable.rs]: Oops, you're right -- I didn't check!

Comment (snim2) [reply src/lib/statetable.rs]: Interesting that that change still passed the unit tests though. Sorry for misleading you!

**Comment by ltratt** [comment]
OK, unfix pushed.

**Comment by snim2** [comment]
OK, time to squash I think!

**Comment by ltratt** [comment]
Squashed.

**Pull request merged**